### PR TITLE
Docs: Coding Guidelines

### PR DESCRIPTION
**Pull request by lmoureaux**
This started as #1321 and ended up being #743.

Sorry for the length, I can split if needed...

### CODE REVIEW

File: docs/Coding/architecture.rst [hunk 1]
```diff
@@ -17,19 +17,47 @@ which is how multiplayer games work.
...
-possible.
+in the client, but the rest of the code base is slowly being converted as well.
+
+Repository Organization
+=======================
+
+The source code is organized in directories at the root of the repository. The client and server share a lot
+of code found in the :file:`common` and :file:`utility` folders. The first one contains code to manage the
+game state, of which both the client and server have a copy, and the second is home to various lower-level
+utilities that don't have an equivalent in Qt or the standard library. Freeciv21 also ships with a couple
```
Comment (jwrober) [inline docs/Coding/architecture.rst:28]: with a couple of...

File: docs/Coding/architecture.rst [hunk 2]
```diff
@@ -17,19 +17,47 @@ which is how multiplayer games work.
...
+external dependencies under :file:`dependencies`.
+
+The client code is grouped in the :file:`client` folder. The server code is located under :file:`server`,
```
Comment (jwrober) [inline docs/Coding/architecture.rst:31]: I would suggest "found" instead of "grouped"

File: docs/Coding/architecture.rst [hunk 3]
```diff
@@ -17,19 +17,47 @@ which is how multiplayer games work.
...
+with the exception of the computer ("AI") players which is under :file:`ai`. The code for other programs
+bundled with Freeciv21, such as the :doc:`Modpack Installer </General/modpack-installer>`, is located under
+:file:`tools`.
+
+All the assets used by client and server are grouped under :file:`data`. This includes among
```
Comment (jwrober) [inline docs/Coding/architecture.rst:36]: used by "the" client and server...

File: docs/Coding/architecture.rst [hunk 4]
```diff
@@ -17,19 +17,47 @@ which is how multiplayer games work.
...
+others :doc:`rulesets </Modding/Rulesets/overview>` and :ref:`tilesets <Modding/index:Tilesets>`.
+Localization files for languages other than English are located under :file:`translations`.
+
+There are a few additional folders that you'll touch less often. The table below describes the complete
```
Comment (jwrober) [inline docs/Coding/architecture.rst:40]: you will

**Review by jwrober — CHANGES_REQUESTED** [review]

File: docs/Coding/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+external dependencies under :file:`dependencies`.
+
+The client code is grouped in the :file:`client` folder. The server code is located under :file:`server`,
+with the exception of the computer ("AI") players which is under :file:`ai`. The code for other programs
+bundled with Freeciv21, such as the :doc:`Modpack Installer </General/modpack-installer>`, is located under
+:file:`tools`.
+
+All the assets used by client and server are grouped under :file:`data`. This includes among
+others :doc:`rulesets </Modding/Rulesets/overview>` and :ref:`tilesets <Modding/index:Tilesets>`.
+Localization files for languages other than English are located under :file:`translations`.
```
Comment (daavko) [inline docs/Coding/architecture.rst:38]: I'd get rid of "for languages other than English", as we do have en-GB in there

File: docs/Coding/guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+-----------------
+
+New code follows the `SPDX standard <https://spdx.dev/ids/>`_:
+
+.. code-block:: cpp
+
+    // SPDX-License-Identifier: GPL-3.0-or-later
+    // SPDX-FileCopyrightText: 2022 Author Name <how-to-contact@example.com>
+
+You do npt need to add your name to the file, but we recommend that you do so. The same information will be
```
Comment (daavko) [inline docs/Coding/guidelines.rst:46]: typo
```suggestion
You do not need to add your name to the file, but we recommend that you do so. The same information will be
```

File: docs/Coding/guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+Header Guards
+-------------
+
+Use ``#pragma once`` to protect headers against multiple inclusion:
+
+.. code-block:: cpp
+
+    #pragma once
+
+This is equivalent to the ancient ``#ifndef``/``#define``/``#endif`` idiom, but also much less error-prone.
```
Comment (daavko) [inline docs/Coding/guidelines.rst:59]:
```suggestion
This is equivalent to the ancient ``#ifndef``/``#define``/``#endif`` idiom, but much less error-prone.
```

File: docs/Coding/guidelines.rst [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+     * Calculates the square of @c x.
+     */
+    int some_internal_function(int x)
+    {
+      return x + IMPORTANT_CONSTANT;
+    }
+
+    } // anonymous namespace
+
+The compiler will generate an arbitrary names for symbols in the anonymous namespace that will not clash with
```
Comment (daavko) [inline docs/Coding/guidelines.rst:185]: Should be either "an arbitrary name" or "arbitrary names" without "an"

File: docs/Coding/guidelines.rst [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+variables, it is especially important for functions taking references (see above).
+
+Functions that do not modify their argument should make them ``const``. Class methods that do not modify the
+object should also be marked ``const``.
+
+
+Use encapsulation
+-----------------
+
+Classes that are more complicated than C-like ``struct`` should not have any public variable. Getters and
```
Comment (daavko) [inline docs/Coding/guidelines.rst:235]:
```suggestion
Classes that are more complicated than C-like ``struct`` should not have any public variables. Getters and
```

File: docs/Coding/architecture.rst [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+heavily on the `Qt Framework <https://doc.qt.io>`_ for platform abstraction. This is currently most visible
+in the client, but the rest of the code base is slowly being converted as well.
+
+Repository Organization
+=======================
+
+The source code is organized in directories at the root of the repository. The client and server share a lot
+of code found in the :file:`common` and :file:`utility` folders. The first one contains code to manage the
+game state, of which both the client and server have a copy, and the second is home to various lower-level
+utilities that don't have an equivalent in Qt or the standard library. Freeciv21 also ships with a couple
```
Comment (daavko) [inline docs/Coding/architecture.rst:28]:
```suggestion
utilities that do not have an equivalent in Qt or the standard library. Freeciv21 also ships with a couple
```

**Review by daavko — COMMENTED** [review]
Looks great in general, found just a couple small things

**Review by jwrober — APPROVED** [review]

**Pull request merged**